crates/fuel-core/src/service/adapters/consensus_parameters_provider.rs

```diff
@@ -0,0 +1,177 @@
+use crate::{
+    database::Database,
+    service::adapters::BlockImporterAdapter,
+};
+use fuel_core_producer::ports::BlockProducerDatabase;
+use fuel_core_services::{
+    stream::BoxStream,
+    RunnableService,
+    RunnableTask,
+    ServiceRunner,
+    SharedMutex,
+    StateWatcher,
+};
+use fuel_core_storage::{
+    not_found,
+    tables::ConsensusParametersVersions,
+    Result as StorageResult,
+    StorageAsRef,
+};
+use fuel_core_txpool::ports::BlockImporter;
+use fuel_core_types::{
+    blockchain::header::ConsensusParametersVersion,
+    fuel_tx::ConsensusParameters,
+    services::block_importer::SharedImportResult,
+};
+use futures::StreamExt;
+use std::{
+    collections::HashMap,
+    sync::Arc,
+};
+
+#[derive(Clone, Debug)]
+pub struct SharedState {
+    latest_consensus_parameters_version: SharedMutex<ConsensusParametersVersion>,
+    consensus_parameters:
+        SharedMutex<HashMap<ConsensusParametersVersion, Arc<ConsensusParameters>>>,
```

Comment:
I would be interested to understand better if we really need `Arc<ConsensusParameters>` here instead of just `ConsensusParameters`:
- If the `SharedState` is already shared, then the contents of `SharedState` are also shared, and the hashmap values can be accessed by cloning this `Arc` and locking `consensus_parameters`
- If `SharedState` owns the `HashMap`,  and the `HashMap` owns the value type, we know that the lifetime of the hashmap is maintained by `SharedState`, which in turn has a lifetime equal to the long running task. We have clear ownership semantics so we probably don't need RC here

Let me know if that sounds possible or if I'm simply misunderstanding anything here.

Reply:
We can't return a reference to the `ConsensusParameters` because it is owned by the `MutexGuard`, so it has a local lifetime. Plus, if we could, it would create a situation where you can return a reference, lock the `HashMap`, and remove the value, making the reference invalid.